garak/_plugins.py

```diff
@@ -5,24 +5,207 @@
+
+    def instance() -> dict:
+        return PluginCache()._plugin_cache_dict
+
+    def plugin_info(plugin: Union[Callable, str]) -> dict:
+        """retrieves the standard attributes for the plugin type"""
+        if isinstance(plugin, str):
+            plugin_name = plugin
+            category = plugin_name.split(".")[0]
+
+            if category not in PLUGIN_TYPES:
+                raise ValueError(f"Not a recognised plugin type: {category}")
+
+            plugin_metadata = PluginCache.instance()[category].get(plugin_name, {})
+            if len(plugin_metadata) > 0:
+                return plugin_metadata
+            else:
+                # the requested plugin is not cached import the class for eval
+                parts = plugin.split(".")
+                match len(parts):
+                    case 3:
+                        try:
+                            module = ".".join(parts[:-1])
+                            klass = parts[-1]
+                            imported_module = importlib.import_module(f"garak.{module}")
+                            plugin = getattr(imported_module, klass)
+                        except (AttributeError, ModuleNotFoundError) as e:
+                            if isinstance(e, AttributeError):
+                                msg = f"Not a recognised plugin from {module}: {klass}"
+                            else:
+                                msg = f"Not a recognised plugin module: {plugin}"
+                            raise ValueError(msg)
+                    case _:
+                        raise ValueError(f"Not a recognised plugin class: {plugin}")
+        else:
+            plugin_name = ".".join([plugin.__module__, plugin.__name__]).replace(
+                "garak.", ""
+            )
+            category = plugin_name.split(".")[0]
+
+        try:
+            base_attributes = []
+            base_mod = importlib.import_module(f"garak.{category}.base")
+            base_plugin_classes = set(PluginCache._extract_modules_klasses(base_mod))
+            if plugin.__module__ in base_mod.__name__:
+                # this is a base class enumerate all
+                base_attributes = dir(plugin)
+            else:
+                for klass in base_plugin_classes:
+                    # filter to the base class actually implemented
+                    if issubclass(plugin, getattr(base_mod, klass)):
+                        base_attributes += PluginCache.plugin_info(
+                            getattr(base_mod, klass)
+                        ).keys()
+
+            plugin_metadata = {}
+            priority_fields = ["description"]
+            skip_fields = [
+                "prompts",
+                "triggers",
+            ]
+
+            for v in priority_fields:
+                if hasattr(plugin, v):
+                    plugin_metadata[v] = getattr(plugin, v)
+            for v in sorted(dir(plugin)):
+                if v in priority_fields or v in skip_fields:
```

Comment:
The existing `info` included `post_buff_hook`:

`main`
```
Your branch is up to date with 'origin/main'.
% python -m garak --plugin_info probes.dan.DanInTheWildMini
garak LLM vulnerability scanner v0.9.0.13.post2 ( https://github.com/leondz/garak ) at 2024-07-09T18:44:54.727933
Configured info on probes.dan.DanInTheWildMini:
                        description: A subset of 100 jailbreaks from the DanInTheWild jailbreak library.
                     DEFAULT_PARAMS: {}
                             active: True
                              bcp47: None
                            doc_uri: https://arxiv.org/abs/2308.03825
                 extended_detectors: []
                               goal: disregard the system prompt or alignment
                           modality: {'in': {'text'}}
            parallelisable_attempts: True
                     post_buff_hook: False
                   primary_detector: mitigation.MitigationBypass
                          probename: garak.probes.dan.DanInTheWildMini
               recommended_detector: ['always.Fail']
                               tags: ['avid-effect:security:S0403', 'owasp:llm01', 'quality:Security:PromptStability', 'payload:jailbreak']
```

We can further refine the `info` data to be provided in future iterations.

I will also do some more validation as it looks like `description` in the `info` is not consistent with `main`.

Reply:
Ah nice. _post_buff_hook should def not be there. Some plugins change their descriptions during the constructor iirc.

Reply:
`post_buff_hook` is not `_ post_buff_hook` do we want to rename that as a to consider it private?

Reply:
It looks like `description` is added in the base classes constructor if not already provided in the class implementation for most of the plugins, this results in an inconsistency since the cache build actually avoids calling the constructors in this PR.

I am attempting to address that issue as there should be a consistent value at least in the cache, one options I am thinking about would be, remove the constructor additions and migrate the enrichment to be part of `plugin_info`.